static/src/javascripts/projects/commercial/modules/dfp/prepare-googletag.ts

```diff
@@ -0,0 +1,169 @@
+	if (storage.session.isAvailable()) {
+		const pageViews =
+			(storage.session.get('gu.commercial.pageViews') as number) || 0;
+		storage.session.set('gu.commercial.pageViews', pageViews + 1);
+	}
+};
+
+const setPageTargeting = (): void => {
+	const pubads = window.googletag?.pubads();
+	if (!pubads) return;
+	// because commercialFeatures may export itself as {} in the event of an exception during construction
+	const targeting = getPageTargeting() as Record<string, any>;
+	Object.keys(targeting).forEach((key) => {
+		pubads.setTargeting(key, targeting[key]);
+	});
+};
+
+const setPublisherProvidedId = (): void => {
+	// Also known as PPID
+	getUserFromApi((user: IdentityUser) => {
+		if (user.privateFields.googleTagId) {
+			window.googletag
+				?.pubads()
+				.setPublisherProvidedId(user.privateFields.googleTagId);
+		}
+	});
+};
+
+export const init = (): Promise<void> => {
+	const setupAdvertising = (): Promise<void> => {
+		// note: fillAdvertSlots isn't synchronous like most buffered cmds, it's a promise. It's put in here to ensure
+		// it strictly follows preceding prepare-googletag work (and the module itself ensures dependencies are
+		// fulfilled), but don't assume fillAdvertSlots is complete when queueing subsequent work using cmd.push
+		window.googletag?.cmd.push(
+			setDfpListeners,
+			// @ts-expect-error - googletag.CommandArray.push() !== Array.push, it accepts multiple fn arguments
+			setPageTargeting,
+			refreshOnResize,
+			() => {
+				void fillAdvertSlots();
+			},
+		);
+
+		onConsentChange((state) => {
+			let canRun = true;
+			if (state.ccpa) {
+				const doNotSell = state.ccpa.doNotSell;
+				// CCPA mode
+				window.googletag?.cmd.push(() => {
+					window.googletag?.pubads().setPrivacySettings({
+						restrictDataProcessing: doNotSell,
+					});
+				});
+				if (!state.ccpa.doNotSell) {
+					window.googletag?.cmd.push(setPublisherProvidedId);
+				}
+			} else {
+				let npaFlag = false;
+				if (state.tcfv2) {
+					// TCFv2 mode
+					npaFlag =
+						Object.keys(state.tcfv2.consents).length === 0 ||
+						Object.values(state.tcfv2.consents).includes(false);
+					canRun = getConsentFor('googletag', state);
+				} else if (state.aus) {
+					// AUS mode
+					// canRun stays true, set NPA flag if consent is retracted
+					npaFlag = !getConsentFor('googletag', state);
+				}
+				window.googletag?.cmd.push(() => {
+					window.googletag
+						?.pubads()
+						.setRequestNonPersonalizedAds(npaFlag ? 1 : 0);
+				});
+				if (!npaFlag) {
+					window.googletag?.cmd.push(setPublisherProvidedId);
+				}
+			}
+			// Prebid will already be loaded, and window.googletag is stubbed in `commercial.js`.
+			// Just load googletag. Prebid will already be loaded, and googletag is already added to the window by Prebid.
+			if (canRun) {
+				void loadScript(
+					(config as {
+						get: (arg: string, defaultValue: any) => string;
+					}).get(
```

Comment:
To go around this, I’ve done https://github.com/guardian/frontend/blob/main/static/src/javascripts/projects/common/modules/identity/api.ts#L11.

Reply:
Fair enough - I'm in two minds about this. On one hand, my approach adds typing but doesn't change the code itself, but on the other hand your approach leads to less repetition and code which is easier to read...

Reply:
Yes, no strong feelings either way!